core/shared/src/main/scala/zio/ZIOConcurrent.scala

```diff
@@ -0,0 +1,153 @@
+        if (s == -maxWaiters) {
+          // full
+          Synchronizer.full
+        } else if (s >= permits) {
+          // enough available permits
+          if (state.compareAndSet(s, s - permits))
+            Task.unit
+          else
+            loop()
+        } else {
+          // not enough permits, add waiter
+          val ns = if (s >= 0) -1 else s - 1
+          if (state.compareAndSet(s, ns)) {
+            val p   = Promise.unsafeMake[Full, Unit](fiberId)
+            val rem = if (s > 0) permits - s else permits
+            waiters.add((new AtomicInteger(rem), p))
+            p.await
+          } else {
+            loop()
+          }
+        }
+      }
+      loop()
+    }
+
+  def release(permits: Int): UIO[Unit] =
+    UIO.effectSuspendTotal {
+      @tailrec def loop(permits: Int): UIO[Unit] =
+        if (permits == 0) {
+          UIO.unit
+        } else {
+          val s = state.get()
+          if (s >= 0) {
+            if (state.compareAndSet(s, s + permits)) {
+              UIO.unit
+            } else {
+              loop(permits)
+            }
+          } else {
+            val w = waiters.peek()
+            if (w != null) {
+              val (waiterPermits, promise) = w
+              val wp                       = waiterPermits.get()
+              if (wp == 0) {
+                // another release call is polling the waiter
+                loop(permits)
+              } else if (wp <= permits) {
+                // enough permits to release the waiter
+                if (waiterPermits.compareAndSet(wp, 0)) {
+                  // at this point, concurrent release calls need to wait
+                  // for the waiter to be polled
+                  // TODO these should never fail but if it does what's the best way to fail because of a bug?
+                  require(waiters.poll() == w)
+                  require(state.incrementAndGet() <= 0)
+                  promise
+                    .succeed(())
+                    .join(release(permits - wp))
+                    .unit
+                    // TODO how do I drop the environment?
+                    .asInstanceOf[UIO[Unit]]
+                } else {
+                  loop(permits)
+                }
+              } else {
+                // not enough permits to release the waiter
+                // so just update the number of permits the waiter
+                // is waiting for
+                if (waiterPermits.compareAndSet(wp, wp - permits)) {
+                  UIO.unit
+                } else {
+                  loop(permits)
+                }
+              }
+            } else {
+              loop(permits)
+            }
+          }
+        }
+      loop(permits)
+    }
+}
+
+object Synchronizer {
+  case class Full() extends Exception
+  private val full = IO.fail(Full())
+
+  def apply(permits: Int, maxWaiters: Option[Int]) =
+    new Synchronizer(permits, maxWaiters.getOrElse(Int.MaxValue))
```

Comment:
Finally this would need a small change to
```scala
for {
  waiters <- UIO(new ConcurrentLinkedQueue[Waiter])
  state   <- Ref.make[State]((permits, waiters))
} yield new Synchronizer(state, maxWaiters.getOrElse(Int.MaxValue))
```